File: Leetcode/0502 IPO.java
class Solution {
    public int findMaximizedCapital(int k, int capital, int[] profits, int[] capitals) {
        int n = profits.length;
        Project[] projects = new Project[n];
        for (int i = 0; i < n; ++i) {
            projects[i] = new Project(capitals[i], profits[i]);
        }
        Arrays.sort(projects, Comparator.comparingInt(Project::capital));

        var profitMaxHeap = new PriorityQueue<Integer>(Comparator.reverseOrder());
        int projectIndex = 0;
        for (int i = 0; i < k; ++i) {
            while (i < n && projects[i].capital <= capital) {
                profitMaxHeap.add(projects[projectIndex++].profit);
            }
            if (profitMaxHeap.isEmpty()) {
                break;
            }
            capital += profitMaxHeap.poll();
        }
        return capital;
    }

    record Project(int capital, int profit) {}
}

File: Leetcode/0502 IPO.py
import heapq


class Solution:
    def findMaximizedCapital(k, capital, profits, capitals):
        projects = list(zip(capitals, profits))
        heapq.heapify(projects)

        profit_max_heap = []
        for _ in range(k):
            while projects and projects[0][0] <= capital:
                _, profit = heapq.heappop(projects)
                # -ve taken because python heap is always min heap
                heapq.heappush(profit_max_heap, -profit)

            if not profit_max_heap:
                break

            # -ve taken because -ve is inserted in max heap
            capital += -heapq.heappop(profit_max_heap)

        return capital
